protos/file.proto

```diff
@@ -19,7 +19,191 @@
+//
+// Buffers:
+//  * There are no buffers
+// Children:
+//  * An array of indices into the dictionary (which are unsigned integers)
+//  * An array of values
+message DictionaryEncoding {
+  // The number of bits in each key
+  uint32 key_width = 1;
+  // The encoding used to encode the keys
+  EncodingType key_encoding = 2;
+  // The encoding used to encode the values
+  EncodingType value_encoding = 3;
+}
+
+// Encoding for variable-length lists of data
+//
+// Buffers:
+//  * There are no buffers
+// Children:
+//  * An array of offsets (which are unsigned integer values)
+//  * An array of values
+message VariableLengthEncoding {
+  // The number of bits in each offset
+  uint32 offset_width = 1;
+  // If true, then lists with length 0 should be considered null values
+  bool empty_is_null = 2;
+  // The encoding used to encode the offsets
+  EncodingType offset_encoding = 3;
+  // The encoding used to encode the values
+  EncodingType value_encoding = 4;
+}
+
+// An encoding describes how an array is physically serialized into buffers
+//
+// Some encodings split an array into multiple arrays.  For example, plain
+// encoding stores an array of values and an array for validity.  Each of those
+// child arrays has its own encoding.  This means that the encoding forms a
+// tree of encodings where the leaves are buffers.
+message EncodingType {
+  oneof type {
+    ConstantEncoding constant = 1;
+    ValueEncoding value = 2;
+    MaskedEncoding masked = 3;
+    DictionaryEncoding dictionary = 4;
+    VariableLengthEncoding variable_length = 5;
+  }
+}
+
+message ColumnMetadata {
+  message Page {
+    // The file offsets of each of the page buffers
+    //
+    // These offsets might point to the column's data section or they
+    // might point to the column's metadata section.
+    //
+    // The number of pages is variable and depends on the encoding.  There
+    // may be zero buffers (e.g. constant encoded data) in which case this
+    // could be empty.
+    repeated uint64 buffer_offsets = 1;
+    // The size (in bytes) of each of the page buffers
+    //
+    // This field will have the same length as `buffer_offsets` and
+    // may be empty.
+    repeated uint64 buffer_sizes = 2;
+    // The file offset to the start of the page data
+    uint64 offset = 3;
+    // The size (in bytes) of the page data
+    uint64 size = 4;
+    // Logical length (e.g. # rows) of the page
+    uint32 length = 5;
+    // The top-level node of the encoding tree used to encode the page
+    EncodingType encoding = 6;
+  }
+  repeated Page pages = 1;
   
+  // Statistics store the min/max/nulls/... statistics of a page.
+  // There is an array for each statistic.  The logical length of the
+  // arrays is equal the the number of pages for the column.  The number of
+  // statistics pages may be smaller than the number of column pages (in most
+  // cases there will only be one page of statistics)
+  message ColumnStatistics {
```

Comment:
Where do these go?

Reply:
The buffers will be in the metadata section.

Reply:
I realized I forgot an actual reference to this message.  I've added that in `ColumnMetadata`:

```
  repeated Page pages = 1;
  ColumnStatistics statistics = 2;
```